src/openslide-vendor-dicom.c

```diff
@@ -0,0 +1,840 @@
+    return false;
+  }
+
+  print_frame(frame);
+
+  return _openslide_jpeg_decode_buffer(dcm_frame_get_value(frame),
+                                       dcm_frame_get_length(frame),
+                                       dest,
+                                       a->base.w, a->base.h,
+                                       err);
+}
+
+static void associated_destroy(struct associated *a) {
+  dicom_file_destroy(a->file);
+  g_slice_free(struct associated, a);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(associated, associated_destroy)
+
+static void openslide_associated_destroy(struct _openslide_associated_image *_img) {
+  struct associated *a = (struct associated *) _img;
+  associated_destroy(a);
+}
+
+static const struct _openslide_associated_image_ops dicom_associated_ops = {
+  .get_argb_data = associated_get_argb_data,
+  .destroy = openslide_associated_destroy
+};
+
+static bool read_whole_file(struct dicom_file *f) {
+  // try to read the rest of the dicom file
+  if (!f->metadata) {
+    f->metadata = dcm_filehandle_read_metadata(NULL, f->filehandle);
+  }
+  if (!f->metadata) {
+    return false;
+  }
+
+  if (!f->bot) {
+    f->bot = dcm_filehandle_read_bot(NULL, f->filehandle, f->metadata);
+  }
+  if (!f->bot) {
+    /* Try to build the BOT instead.
+     */
+    f->bot = dcm_filehandle_build_bot(NULL, f->filehandle, f->metadata);
+  }
+  if (!f->bot) {
+    return false;
+  }
+
+  return true;
+}
+
+static struct associated *associated_new(struct dicom_file *f) {
+  // we use BOT to get the label and overview frames
+  if (!read_whole_file(f)) {
+    return NULL;
+  }
+
+  // ImageType must be one of the combinations we accept
+  if (!is_type(f, &associated_types)) {
+    return NULL;
+  }
+
+  const char *name;
+  if (!get_tag_str(f->metadata, "ImageType", 2, &name)) {
+    return NULL;
+  }
+
+  g_autoptr(associated) a = g_slice_new0(struct associated);
+
+  if (!get_tag_int(f->metadata, TotalPixelMatrixColumns, &a->base.w) ||
+    !get_tag_int(f->metadata, TotalPixelMatrixRows, &a->base.h)) {
+    return NULL;
+  }
+
+  // set this once we know we will succeed ... this passes ownership to the
+  // level
+  a->file = f;
+  a->name = name;
+  a->base.ops = &dicom_associated_ops;
+
+  g_debug("associated_new: %s", a->name);
+
+  return g_steal_pointer(&a);
+}
+
+static struct dicom_level *level_new(struct dicom_file *f) {
+  if (!read_whole_file(f)) {
+    return NULL;
+  }
+
+  g_autoptr(dicom_level) l = g_slice_new0(struct dicom_level);
+
+  if (!get_tag_int(f->metadata, TotalPixelMatrixColumns, &l->base.w) ||
+    !get_tag_int(f->metadata, TotalPixelMatrixRows, &l->base.h) ||
+    !get_tag_int(f->metadata, Columns, &l->base.tile_w) ||
+    !get_tag_int(f->metadata, Rows, &l->base.tile_h)) {
+    return NULL;
+  }
+
+  // ImageType must be one of the combinations we accept
+  if (!is_type(f, &level_types)) {
+    return NULL;
+  }
+
+  // we only allow square tiles
+  if (l->base.tile_w != l->base.tile_h) {
+    return NULL;
```

Comment:
Currently, when this function rejects a level, we just continue looking for other levels.  That makes sense when ignoring truly optional data (e.g. maybe an associated image we don't understand), but OpenSlide generally tries to reject slides it doesn't understand rather than silently producing incomplete data or garbage.  Should some of these tests be converted into hard failures?

Reply:
DICOM is extremely flexible and extensible, even more than tiff, which is a blessing and a curse. I think DICOM viewers are expected to display what they can.